game/ui/consolewidget.cpp

```diff
@@ -137,36 +137,64 @@ void ConsoleWidget::keyDownEvent(KeyEvent& e) {
     close();
     }
 
+  if(e.key==Event::K_Left) {
+    if (cursPos>0)
+      cursPos--;
+    return;
+    }
+  if(e.key==Event::K_Right) {
+    if(cursPos<log.back().size())
+      cursPos++;
+    return;
+    }
   if(e.key==Event::K_Up) {
     histPos++;
-    if(histPos<cmdHist.size())
-      log.back() = cmdHist[cmdHist.size()-1-histPos]; else
-      histPos    = size_t(cmdHist.size()-1);
+    if(histPos<cmdHist.size()) {
+      if(histPos==0)
+        currCmd = log.back();
+      log.back() = cmdHist[cmdHist.size()-1-histPos];
+      cursPos    = log.back().size();
+      }
+    else {
+      histPos = size_t(cmdHist.size()-1);
+      }
     return;
     }
   if(e.key==Event::K_Down) {
     if(histPos==size_t(-1))
       return;
     histPos--;
     if(histPos<cmdHist.size())
-      log.back() = cmdHist[cmdHist.size()-1-histPos]; else
-      log.back() = "";
+      log.back() = cmdHist[cmdHist.size()-1-histPos];
+    else
+      log.back() = currCmd;
+    cursPos = log.back().size();
     return;
     }
 
   if(e.key==Event::K_C && (e.modifier&Event::M_Command)==Event::M_Command) {
     log.emplace_back("");
+    cursPos = 0;
+    currCmd = "";
+    histPos = size_t(-1);
     return;
     }
 
   if(e.key==Event::K_Back) {
-    if(log.back().size()>0)
-      log.back().pop_back();
+    if(log.back().size()>0 && cursPos>0) {
```

Comment:
Maybe something like 
`if(0<curssPos && curssPos<=log.back().size())`
would be more clean way to express this condition?